feat(curriculum): add web standards model lectures (#67270)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/lecture-understanding-the-web-standards-model/69fc757399f4066e674a5160.md

title: What's the Process for Creating Web Standards?
challengeType: 19
dashedName: whats-the-process-for-creating-web-standards
---

# --description--

The process of creating web standards varies by body. The W3C runs formal working groups with structured review stages. WHATWG operates more like an open-source project. TC39 holds in-person meetings every two months where delegates vote on whether proposals advance. The Khronos Group gates participation by membership tier.

However, underneath those differences, the process of turning an idea into a web standard follows an observable pattern.

1. ## Identifying a need
    Standards start with friction, like developers hitting a brick wall, browsers implementing the same behavior in incompatible ways, or a new capability the platform simply doesn't support yet. Someone, somewhere, gets frustrated enough to write it down. That someone can be anyone; a developer, a large company with a product need, or an engineer inside a browser team. What matters at this stage is not authority, but evidence and questions on whether the problem is real, and whether enough people are running into it to justify the web platform's attention.

> 1. ## Proposal and early discussion

Reviewer: ```suggestion
1. **Proposal and early discussion**
```

    The idea gets written up informally through a GitHub issue, a short document, or an explainer that describes the problem and sketches a possible solution. At this stage, the goal is not to produce a finished specification but to find out whether the problem resonates. Do other developers recognize it? Are browser vendors interested? Is there an obvious solution, or several competing ones? In TC39, this early phase corresponds to Stage 0 and Stage 1. In the W3C, it often happens informally in a Community Group before a formal Working Group is chartered. The WHATWG handles it through open GitHub discussions on its core repositories. The venue differs, but the function is the same.

> 1. ## Writing the specification

Reviewer: ```suggestion
1. **Writing the specification**
```

    If the proposal gains traction, someone or a small team has to write the actual specification, or spec. A spec is not a blog post or a tutorial, but a precise, formal document describing exactly how a feature must behave in every possible situation. Edge cases, error conditions, interactions with existing features, and security implications all have to be thought through and written down.

> 1. ## Review, objection, and iteration

Reviewer: ```suggestion
1. **Review, objection, and iteration**
```

    The draft is published and reviewed publicly, and by the other browser vendors. This is where most of the real work happens. Objections are raised, and the spec is revised. A feature that one major browser engine refuses to implement will rarely become a standard, regardless of its technical merits. Consensus-building at this stage can take a long time. Some proposals stall for years while vendors negotiate over design details. Others die entirely when it becomes clear that no agreement is reachable.

> 1. ## Implementation and testing

Reviewer: ```suggestion
1. **Implementation and testing**
```

>     Once a spec is stable enough, browser vendors begin implementing it. This usually starts with a feature flag first, so it can be tested without affecting existing users. Implementation often reveals problems that the spec didn't anticipate, and then sending things back to the editors for clarification. A shared test suite verifies that every implementation actually behaves the same way. If Chrome, Firefox, and Safari all pass the same tests, the spec is working. If they diverge, either the spec is unclear, or one of the implementations has a bug.

Reviewer: ```suggestion
1. ## Implementation and testing
    Once a spec is stable enough, browser vendors begin implementing it. This usually starts with a feature flag first, so it can be tested without affecting existing users. Implementation often reveals problems that the spec didn't anticipate, which then get sent back to the editors for clarification. A shared test suite verifies that every implementation actually behaves the same way. If Chrome, Firefox, and Safari all pass the same tests, the spec is working. If they diverge, either the spec is unclear, or one of the implementations has a bug.
```

---
id: 69fc757399f4066e674a5160
title: What's the Process for Creating Web Standards?
challengeType: 19
dashedName: whats-the-process-for-creating-web-standards
---

# --description--

The process of creating web standards varies by body. The W3C runs formal working groups with structured review stages. WHATWG operates more like an open-source project. TC39 holds in-person meetings every two months where delegates vote on whether proposals advance. The Khronos Group gates participation by membership tier.

However, underneath those differences, the process of turning an idea into a web standard follows an observable pattern.

1. ## Identifying a need

>     Standards start with friction, like developers hitting a brick wall, browsers implementing the same behavior in incompatible ways, or a new capability the platform simply doesn't support yet. Someone, somewhere, gets frustrated enough to write it down. That someone can be anyone, a developer, a large company with a product need, or an engineer inside a browser team. What matters at this stage is not authority, but evidence and questions on whether the problem is real, and whether enough people are running into it to justify the web platform's attention.

Reviewer 2: ```suggestion
    Standards start with friction, like developers hitting a brick wall, browsers implementing the same behavior in incompatible ways, or a new capability the platform simply doesn't support yet. Someone, somewhere, gets frustrated enough to write it down. That someone can be anyone; a developer, a large company with a product need, or an engineer inside a browser team. What matters at this stage is not authority, but evidence and questions on whether the problem is real, and whether enough people are running into it to justify the web platform's attention.
```

1. ## Proposal and early discussion

>     The idea gets written up informally through a GitHub issue, a short document, or an explainer that describes the problem and sketches a possible solution. At this stage, the goal is not to produce a finished spec but to find out whether the problem resonates. Do other developers recognize it? Are browser vendors interested? Is there an obvious solution, or several competing ones? In TC39, this early phase corresponds to Stage 0 and Stage 1. In the W3C, it often happens informally in a Community Group before a formal Working Group is chartered. The WHATWG handles it through open GitHub discussions on its core repositories. The venue differs, but the function is the same.

Reviewer 2: ```suggestion
    The idea gets written up informally through a GitHub issue, a short document, or an explainer that describes the problem and sketches a possible solution. At this stage, the goal is not to produce a finished specification but to find out whether the problem resonates. Do other developers recognize it? Are browser vendors interested? Is there an obvious solution, or several competing ones? In TC39, this early phase corresponds to Stage 0 and Stage 1. In the W3C, it often happens informally in a Community Group before a formal Working Group is chartered. The WHATWG handles it through open GitHub discussions on its core repositories. The venue differs, but the function is the same.
```

1. ## Writing the specification

>     If the proposal gains traction, someone or a small team has to write the actual specification. A spec is not a blog post or a tutorial, but a precise, formal document describing exactly how a feature must behave in every possible situation. Edge cases, error conditions, interactions with existing features, and security implications all have to be thought through and written down.

Reviewer 2: ```suggestion
    If the proposal gains traction, someone or a small team has to write the actual specification, or spec. A spec is not a blog post or a tutorial, but a precise, formal document describing exactly how a feature must behave in every possible situation. Edge cases, error conditions, interactions with existing features, and security implications all have to be thought through and written down.
```

1. ## Review, objection, and iteration

>     The draft is published and reviewed publicly, and by the other browser vendors. This is where most of the real work happens. Objections are raised, and the spec is revised. A feature that one major browser engine refuses to implement will rarely become a standard, regardless of its technical merits. Consensus-building at this stage can take a long time. Some proposals stall for years while vendors negotiate over design details. Others die entirely when it becomes clear that no agreement is reachable. The ones that survive are, by definition, the ones that enough powerful stakeholders could live with, which is not always the same as the best technical solution.

Reviewer 2: ```suggestion
    The draft is published and reviewed publicly, and by the other browser vendors. This is where most of the real work happens. Objections are raised, and the spec is revised. A feature that one major browser engine refuses to implement will rarely become a standard, regardless of its technical merits. Consensus-building at this stage can take a long time. Some proposals stall for years while vendors negotiate over design details. Others die entirely when it becomes clear that no agreement is reachable.
```

I worry about this sentence being seen as a "mission" or such~

1. ## Implementation and testing

>     Once a spec is stable enough, browser vendors begin implementing it. This usually starts with a feature flag first, so it can be tested without affecting existing users. Implementation often reveals problems that the spec didn't anticipate, and then sending things back to the editors for clarification. A shared test suite verifies that every implementation actually behaves the same way. If Chrome, Firefox, and Safari all pass the same tests, the spec is working. If they diverge, either the spec is unclear, or one of the implementations has a bug. Both happen more often than anyone would like.

Reviewer 2: ```suggestion
    Once a spec is stable enough, browser vendors begin implementing it. This usually starts with a feature flag first, so it can be tested without affecting existing users. Implementation often reveals problems that the spec didn't anticipate, and then sending things back to the editors for clarification. A shared test suite verifies that every implementation actually behaves the same way. If Chrome, Firefox, and Safari all pass the same tests, the spec is working. If they diverge, either the spec is unclear, or one of the implementations has a bug.
```

Unnecessary and subjective I think~

> 

Reviewer 2: ```suggestion
```

# --questions--

## --text--

What is typically the first step in turning an idea into a web standard?

## --answers--

Writing the full specification.

### --feedback--

Think about what motivates people to start the standards process in the first place.

---

Getting approval from a major browser vendor.

### --feedback--

Think about what motivates people to start the standards process in the first place.

---

> Identifying a real need of developers.

Reviewer 2: ```suggestion
Identifying a real need experienced by developers.
```

---

Submitting a formal patent application.

### --feedback--

Think about what motivates people to start the standards process in the first place.

## --video-solution--

3

## --text--

What happens during the review and iteration stage of creating a web standard?

## --answers--

> The standard is automatically approved once written

Reviewer 2: ```suggestion
The standard is automatically approved once written.
```

### --feedback--

Think about how disagreements get resolved before a feature ships in all browsers.

---

Browser vendors raise objections, and the spec is revised until consensus is reached.

---

Only the original author can make changes.

### --feedback--

Think about how disagreements get resolved before a feature ships in all browsers.

---

> The standard is sent to a lawyer for copyright approval

Reviewer 2: ```suggestion
The standard is sent to a lawyer for copyright approval.
```

### --feedback--

Think about how disagreements get resolved before a feature ships in all browsers.

## --video-solution--

2

## --text--

Why is a shared test suite important during the implementation stage of a web standard?

## --answers--

> To replace the need for a written specification

Reviewer 2: ```suggestion
To replace the need for a written specification.
```

### --feedback--

Think about what ensures Chrome, Firefox, and Safari act identically.

---

> To verify that all browser implementations behave the same way

Reviewer 2: ```suggestion
To verify that all browser implementations behave the same way.
```

---

> To rank browser performance for marketing purposes

Reviewer 2: ```suggestion
To rank browser performance for marketing purposes.
```

### --feedback--

Think about what ensures Chrome, Firefox, and Safari act identically.

---

> To prevent developers from using the new feature

Reviewer 2: ```suggestion
To prevent developers from using the new feature.
```
